llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -18602,6 +18602,30 @@ static SDValue performBuildShuffleExtendCombine(SDValue BV, SelectionDAG &DAG) {
     SeenZExtOrSExt = true;
   }
 
+  // Avoid the said use of vector SExt/ZExt in case all vector elements are
+  // consumed and each shuffle's mask uses same index, in order to permit use of
+  // indexed OP (e.g. MLA, MUL) variants
+  EVT ExtendType = Extend->getValueType(0);
+  if (ExtendType.isVector() && !ExtendType.isScalableVT()) {
+    const int NumElements = ExtendType.getVectorNumElements();
+    SmallBitVector UsedElements(NumElements, false);
+    for (SDNode *User : Extend.getNode()->users()) {
+      if (User->getOpcode() == ISD::VECTOR_SHUFFLE &&
+          User->getOperand(0) == Extend) {
+        ArrayRef<int> Mask = cast<ShuffleVectorSDNode>(User)->getMask();
+        const int Idx = Mask[0];
+        if (Idx >= NumElements)
+          continue;
+        if (llvm::all_of(Mask, [Idx](int M) { return M == Idx; }))
+          UsedElements.set(Idx);
+        else
+          break; // early loop exit to help performance
+      }
```

Comment:
Early exit are generally preferred in LLVM to reduce nesting level and simplify the code (https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code). E.g. something like 
```suggestion
       if (User->getOpcode() != ISD::VECTOR_SHUFFLE ||  
          User->getOperand(0) != Extend)
        break;

      ArrayRef<int> Mask = cast<ShuffleVectorSDNode>(User)->getMask();
      const int Idx = Mask[0];
      if (Idx >= NumElements || !all_of(Mask, [Idx](int M) { return M == Idx; }))
        break;
      UsedElements.set(Idx);
```

Reply:
Fair point.

Note however, a `break` in the first if-statement would skip some good cases. For example, when there is a non-shuffle user of the Extend. A `continue` should be used here.

I pushed a fix for that as well as one more test-case called `ext_shuffle_v4i16_v4i32_extra_user` for that said example.